### DISCUSSION TITLE
Move-PnpFile resulting in Access Denied

### DISCUSSION THREAD

**Discussion by wil16098** [pnp/powershell#1447] [Q&A]
Hello. Thank you for your hard work on this project. (especially @veronicageek - love the youtube channel). I am encountering the following errors when attempting to move the contents of a departed users OneDrive to a site library called 'PriorUsersOneDrive'. Each folder in PriorUsersOneDrive is named for the MailNickname of the user. Example: /sites/PriorUsersOneDrive/testuser if the email is testuser@mydomain.com. 
I am able to connect just fine. I can Find-PnPFile in both sites just fine. I can Add-PnPFile to both sites just fine. I am a Global Admin and Site Collection Admin for both sites and have added code to (I hope) reinforce that ($GlobalAdminEmail entered previously and MFA authenticating just fine). I am using PnP.Powershell (1.8.0). Here is my code. I am pretty confident I don't need all the connect statements, but was trying to make 100% sure I was hitting the right spot, and removing that bit doesn't help.
I have also included my resulting output and thank you in advance for any assistance on why I am receiving this access denied error. I am 100% certain it is user error on my part and appreciate any insights in advance :):

```    
    #SharePoint Online via SPO and PnP
        Connect-SPOService -Url https://$ClientOrg-admin.sharepoint.com
        Connect-PnPOnline -Url https://$ClientOrg.sharepoint.com -Interactive
    #Create the user's (name=email handle) folder in the Prior User's OneDrive Document Library using PnP.PowerShell (if it does not already exist)
        $PUODSite=Get-SPOSite -Identity https://$ClientOrg.sharepoint.com/sites/PriorUsersOneDrive
        Connect-PnPOnline -Url $PUODSite.Url -Interactive
        Resolve-PnPFolder -SiteRelativePath "Shared Documents/$($AADUserID.MailNickName)"
    #Make sure the GlobalAdmin is a Site Collection Admin, allowing you to do operations on the user's OneDrive
        $SourceAccount= Get-PnPUserProfileProperty -Account $userUPN
        $DestinationFolder= "sites/PriorUsersOneDrive/Shared Documents/$($AADUserID.MailNickName)"
        Set-SPOUser -Site $SourceAccount.PersonalUrl -LoginName $GlobalAdminEmail -IsSiteCollectionAdmin $true
        Set-SPOUser -Site $PUODSite.Url -LoginName $GlobalAdminEmail -IsSiteCollectionAdmin $true
    #Move the user's OneDrive files to the newly created Prior Users OneDrive folder named for them after connecting to the user's OneDrive folder        
        Connect-PnPOnline -Url $SourceAccount.PersonalUrl -Interactive
        Move-PnPFile -SourceUrl $SourceAccount.PersonalUrl -TargetUrl $DestinationFolder -AllowSchemaMismatch -AllowSmallerVersionLimitOnDestination -Force -Overwrite
```

Error text:
```
Move-PnPFile : Access denied.
At line:12 char:9
+         Move-PnPFile -SourceUrl $SourceAccount.PersonalUrl -TargetUrl ...
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : WriteError: (:) [Move-PnPFile], ServerUnauthorizedAccessException
    + FullyQualifiedErrorId : EXCEPTION,PnP.PowerShell.Commands.Files.MoveFile
```
Again, finding and adding is working just fine:
```
Find-PnPFile -Folder $SourceAccount.PersonalUrl -match *.docx
Name                  Type Items/Size Last Modified         
----                  ---- ---------- -------------         
Test Document 1 .docx File            12/14/2021 10:54:17 PM
Document.docx         File            12/14/2021 10:51:33 PM
Test Document 2.docx  File            12/14/2021 10:54:40 PM
Test Document 3.docx  File            12/14/2021 10:54:55 PM

PS C:\WINDOWS\system32> Add-PnPFolder -Name "Folder1" -Folder "Documents"
Name    Type   Items/Size Last Modified        
----    ----   ---------- -------------        
Folder1 Folder          0 12/15/2021 3:42:39 AM
```
Both of the above were run against the user's current OneDrive successfully

**Comment by AndersRask** [comment]
Your TargetUrl is a relative path. It will need an absolute path if you want to move it to another site, so it probably tries to move it to a non-existent subsite, which somewhat explains the error (errors in SharePoint are usually confusing and vague):

`-TargetUrl <String>
        Site or server relative URL of a document library where to move the file or folder to.`

Try 
`$DestinationFolder= "/sites/PriorUsersOneDrive/Shared Documents/$($AADUserID.MailNickName)"`

It is also mentioned in example:
```
Move-PnPFile -SourceUrl "Shared Documents/Document.docx" -TargetUrl "/sites/otherproject/Shared Documents" -OverwriteIfAlreadyExists -AllowSchemaMismatch
    -AllowSmallerVersionLimitOnDestination

    Moves a file named Document.docx located in the document library named "Shared Documents" in the current site to the document library named "Shared Documents" in
    another site collection "otherproject" allowing it to overwrite an existing file Document.docx in the destination, allowing the fields to be different on the
    destination document library from the source document library and allowing a lower document version limit on the destination compared to the source.
    -------------------------- EXAMPLE 4 --------------------------

    Move-PnPFile -SourceUrl "/sites/project/Shared Documents/Archive" -TargetUrl "/sites/archive/Project" -AllowSchemaMismatch -AllowSmallerVersionLimitOnDestination

    Moves a folder named Archive located in the document library named "Shared Documents" in the current site to the document library named "Project" in another site
    collection "archive" not allowing it to overwrite an existing folder named "Archive" in the destination, allowing the fields to be different on the destination
    document library from the source document library and allowing a lower document version limit on the destination compared to the source.
```

**Comment by wil16098** [comment]
Thank you.

If I understand correctly, you have added a leading '/' in front of the $DestinationFolder path. I appreciate that note and tried this syntax based on the example in the documentation, but (retried it again just now) I got this:
```
Move-PnPFile : The remote server returned an error: (404) Not Found.
At line:12 char:9
+         Move-PnPFile -SourceUrl $SourceAccount.PersonalUrl -TargetUrl ...
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : WriteError: (:) [Move-PnPFile], WebException
    + FullyQualifiedErrorId : EXCEPTION,PnP.PowerShell.Commands.Files.MoveFile
```
That seemed to be some sort of pathing error and so I tried it without the '/' and got access denied... but it sounded like without the '/' it was maybe at least finding the correct path?
Thank you again so much for trying to help me with this!

**Comment by AndersRask** [comment]
If you read the examples carefully [here ](https://pnp.github.io/powershell/cmdlets/Move-PnPFile.html) you will notice that you need "/" in front (absolute URL) if you want to move to another site collection: -TargetUrl "/sites/otherproject/Shared Documents". 
If you omit it (relative URL) it will, depending on syntax in URL, try to move it to either a subsite (-TargetUrl "SubSite2/Shared Documents"), another library (-TargetUrl "Archive/Document2.docx"). In other words, you do net a slash :)

Looking at your code, it actually seems to be the SourceUrl that is the culprit. You are just pointing to the root of the users MySite. You need to select the appropriate document libraries and either iterate over files where you move, or maybe use RootFolder. This is also why you get 404: PowerShell does not know where to look for documents :)

Also it is super confusing that you use both SPO and PnP. If you need to connect to multiple sites at once, just dedicate the connection using -ReturnConnection parameter and then use -Connection parameter when calling cmdlets

$siteA = Connect-PnPOnline -ReturnConnection -Url "https://contoso.sharepoint.com/sites/siteA"
$siteB = Connect-PnPOnline -ReturnConnection -Url "https://contoso.sharepoint.com/sites/siteB"

Get-PnPListItem -List "Documents" -Connection $siteA
Get-PnPListItem -List "Documents" -Connection $siteB

**Comment by wil16098** [comment]
Thank you very much for the code examples! I implemented as follows:
```
$ClientOrg= [the 'clientorg' name in - 'clientorg.onmicrosoft.com']
$UserUPN= [full email account of user in the format - 'testuser@clientorg.com']
$AADUserID = Get-AzureADUser -ObjectId "$UserUPN"

$SourceAccount= Get-PnPUserProfileProperty -Account $userUPN
$DestinationFolder= "https://$ClientOrg.sharepoint.com/sites/PriorUsersOneDrive/Shared Documents/$($AADUserID.MailNickName)"

$siteA = Connect-PnPOnline -ReturnConnection -Url $SourceAccount.PersonalUrl -Interactive
$siteB = Connect-PnPOnline -ReturnConnection -Url $DestinationFolder -Interactive

Get-PnPListItem -List "Documents" -Connection $siteA
Get-PnPListItem -List "Documents" -Connection $siteB
```
I am able to connect to $siteA and $siteB without error.
For my SourceAccount command:
```
Get-PnPListItem -List "Documents" -Connection $siteA
```
I got:
```
Id    Title                                              GUID                                                                                                                                                         
--    -----                                              ----                                                                                                                                                                                                                                                                             
5     Test Folder 2                                      b1e7a2e9-be68-41cd-a20f-474c86c7214e                                                                                                                                                                                                                                                
9     Folder1                                            3e755909-622a-436d-aab0-fb6b744f7f90                                                                                                                         
...                                                                                                                      
7                                                        597cf3b4-3e39-4d7b-91c9-166792c387b0 
```
As expected.
However, my DestinationFolder command:
```
Get-PnPListItem -List "Documents" -Connection $siteB
```
Returns:
```
Get-PnPListItem : Cannot contact site at the specified URL https://[$ClientOrg].sharepoint.com/sites/priorusersonedrive/shared documents/testuser. There is no Web named "/sites/priorusersonedrive/shared 
documents/testuser/_vti_bin/sites.asmx".
At line:1 char:1
+ Get-PnPListItem -List "Documents" -Connection $siteB
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : WriteError: (:) [Get-PnPListItem], ClientRequestException
    + FullyQualifiedErrorId : EXCEPTION,PnP.PowerShell.Commands.Lists.GetListItem
```
I have tried several other URL combinations to connect to the testuser folder within PriorUsersOneDrive. All of them give errors when trying to connect.

The full URL in $DestinationFolder above allows me to connect just fine. However, I cannot list items once connected (there is a test doc and a sample doc within $Destination Folder). Instead, I receive the above error.

Interestingly, the following:
```
$DestinationFolder= "https://$ClientOrg.sharepoint.com/sites/PriorUsersOneDrive"
$siteB = Connect-PnPOnline -ReturnConnection -Url $DestinationFolder -Interactive
Get-PnPListItem -List "Documents" -Connection $siteB
```
Returns:
```
Id    Title                                              GUID                                                                                                                                                         
--    -----                                              ----                                                                                                                                                         
1                                                        54005400-3b09-4b5c-a2b0-e88b38a88e14                                                                                                                         
2                                                        76b25996-e4dd-4379-8232-998b6abc182f                                                                                                                         
...                                                     d8ac2ee6-617a-4941-b62b-626afd3d3213                                                                                                                         
220   testuser                                           8bfcb139-3e8d-4bf9-bffb-4c57702d4197                                                                                                                         
221                                                      01f9aebe-cbec-482b-9ef5-96912a2e6bf0                                                                                                                         
```
As expected (including the testuser folder).

This leads me to believe my connection to $SourceAccount.PersonalUrl is good, but I am having trouble with $DestinationFolder? Any suggestions as to why? And any suggestions as to how I accomplish the intended move of all contents from siteA to the specified folder in siteB?
It appears I can connect to and list items in "https://$ClientOrg.sharepoint.com/sites/PriorUsersOneDrive"
I can connect to any specific subfolder under PriorUsersOneDrive.
However, I am unable to Get-PnPListItem for any specific subfolder under PriorUsersOneDrive.

Again, thank you for looking at my issue so closely!

**Reply by AndersRask** [reply]
I don't think you read my answer very closely, so i'll try again :)

**Looking at your code, it actually seems to be the SourceUrl that is the culprit. You are just pointing to the root of the users MySite. You need to select the appropriate document libraries and either iterate over files where you move, or maybe use RootFolder. This is also why you get 404: PowerShell does not know where to look for documents :)**

I'm not arguing if the url is good or bad, it just isn't pointing to a file or folder. Pointing to the root of the personal site does not do anything (other than giving you the 404 you see).

You need to either get RootFolder (havent tested if cmdlet works with a hierarchy of subfolders) loop over files/folders in Documents library calling Move-PnPFile for each item

**Reply by wil16098** [reply]
Thank you,

I admit I thought the ability to perform Get-PnPFile, Add-PnPFile, etc. operations on $SourceAccount.PersonalUrl successfully, but not perform these operations on the $DestinationFolder at all, indicated the problem was with $DestinationFolder.

It appears I am wrong and there is something here I do not understand.
How should I code this operation in order to successfully move files from the User's OneDrive to a target subfolder in PriorUsersOneDrive as described above?

If it helps any, I am reading as closely and clearly as I am able to with my current skill level :)

**Comment by wil16098** [comment]
I believe I have determined that what I am trying to do (move files and folders directly from a user's OneDrive to a company SharePoint site) cannot work using Move-PnPFile. 

TLDR - Move-PnPFile cannot support two separate connections at the same time. For this use case, two separate connections are required. One Connection for the user's OneDrive and one connection for the target SharePoint site. Therefore, a straight transfer between a user's OneDrive for Business and a company SharePoint site using Move-PnPFile is not possible. 
My current anticipated workaround will be to grab all the Source objects using a connection A, store locally, then upload to the Target using a different Connection B.

**I am upgrading this thread to an Issue to see if I can get verification or a better workaround.**

The OneDrive for Business folder exists at https://$ClientOrg-my.sharepoint.com/personal/testuser_clientdomain_com/Documents
The Target exists at https://$ClientOrg.sharepoint.com/sites/PriorUsersOneDrive/testuser

Those two domains appear to require two separate connections. So the following, using two separate connections, one for each command, works:
```
$siteA = Connect-PnPOnline -ReturnConnection -Url $SourceAccount.PersonalUrl -Interactive
$siteB = Connect-PnPOnline -ReturnConnection -Url $DestinationFolder -Interactive
Get-PnPListItem -List "Documents" -Connection $siteA
Get-PnPListItem -List "Documents" -Connection $siteB
```
However, Move-PnpFile has and uses only one connection parameter. Therefore, depending on which connection you are using, you will get 'access denied' or '404 not found', etc. based on the connection used. I believe this is because they are different domains:
clientorg-my.sharepoint.com
vs
clientorg.sharepoint.com
and Move-PnPFile will not take an entire url spelled out for each parameter and/or a connection to both domains.
So this does not work:
```
Connect-PnPOnline -Url $SourceAccount.PersonalUrl -Interactive
Move-PnPFile -SourceUrl "/sites/PriorUsersOneDrive/Shared Documents" -TargetUrl "/sites/Test/Shared Documents/TestXfer/testuser"
```
But this does:
```
Connect-PnPOnline -Url $PUODSite.Url -Interactive
Move-PnPFile -SourceUrl "/sites/PriorUsersOneDrive/Shared Documents" -TargetUrl "/sites/Test/Shared Documents/TestXfer/testuser"
```
I would very much love to have someone confirm my suspicion is correct, confirm the proposed workaround is correct, or tell me an elegant way to solve this puzzle. @veronicageek or anyone?

**Reply by AndersRask** [reply]
Hi,

i think you are on to something regarding MySite and "normal" site, as it works fine between two site collections in my tests, so I spent some time looking into the code that is executed.

```csharp
if (!SourceUrl.StartsWith("/"))
{
    SourceUrl = UrlUtility.Combine(webServerRelativeUrl, SourceUrl);
}
if (!TargetUrl.StartsWith("/"))
{
    TargetUrl = UrlUtility.Combine(webServerRelativeUrl, TargetUrl);
}
```

The way the fully qualified target and source urls are generated are by using UrlUtility.Combine() from OfficeDevPnP.Core.Utilities. I think it basically concatenates the  path and relative path for both target and source. And since these are different because MySite and "normal" site are in two different Web Applications it cannot find the site giving you a 404

I think it is safe to say that even this is probably a bug (or at least something that should be specified in the cmdlet that it only works within same WebApp) but not something that I think is easily fixed.

As a workaround you could download the files to a temporary location using `Get-PnPFile` and then copy them to the Archive site, then deleting the site or the files.

You would have to take care of folders as well, but a simple sample with files only could be something like

```powershell
$localFolder = "c:\archive\user1"
$files = Get-PnPFolderItem -FolderSiteRelativeUrl "Documents" -ItemType File  
foreach ($file in $files) {  
    Get-PnPFile -Url $file.ServerRelativeUrl -Path $localFolder -FileName $file.Name - AsFile  
} 
```

hope this helps

kind regards
Anders
